src/alfasim_sdk/result_reader/aggregator.py

```diff
@@ -205,18 +295,52 @@ def map_data(
                 for key, data in gsa_metadata.items()
             }
 
-        with open_result_file(
-            result_directory, result_filename="result"
-        ) as result_file:
-            if not result_file:
-                return cls.empty(result_directory=result_directory)
+        return read_results_file(
+            result_directory=result_directory,
+            metadata_class=cls,
+            map_data=map_data,
+            meta_data_attrs=GLOBAL_SENSITIVITY_ANALYSIS_GROUP_NAME,
+        )
 
-            loaded_metadata = json.loads(
-                result_file[META_GROUP_NAME].attrs["global_sensitivity_analysis"]
-            )
-            return cls(
-                gsa_items=map_data(loaded_metadata), result_directory=result_directory
-            )
+
+UQMetadataClass = Union[
+    GlobalSensitivityAnalysisMetadata, UncertaintyPropagationAnalysesMetaData
+]
+
+
+@attr.s(frozen=True)
+class UPResult:
+    """
+    Holder for each uncertainty propagation result.
+    """
+
+    category: str = attr.ib(
+        validator=attr.validators.optional(attr.validators.instance_of(str))
+    )
+    unit: str = attr.ib(
+        validator=attr.validators.optional(attr.validators.instance_of(str))
+    )
+    realization_output: List = attr.ib(default=attr.Factory(List))
+    std_result: np.ndarray = attr.ib(default=attr.Factory(lambda: np.array([])))
+    mean_result: np.ndarray = attr.ib(default=attr.Factory(lambda: np.array([])))
+
+
+def read_results_file(
+    result_directory: Path,
+    metadata_class: Type[UQMetadataClass],
+    map_data: Callable,
+    meta_data_attrs: str,
+) -> UQMetadataClass:
```

Comment:
See if this works:

```suggestion
MetadataClassType = TypeVar("MetadataClassType", bound=UQMetadataClass)

def read_results_file(
    result_directory: Path,
    metadata_class: Type[MetadataClassType],
    map_data: Callable,
    meta_data_attrs: str,
) -> MetadataClassType:
```

This will provide the correct type narrowing.